ENIGMAsystem/SHELL/Audio_Systems/DirectSound/DSbasic.cpp

```diff
@@ -195,108 +192,211 @@ namespace enigma_user
+  if (!fdata) return -1;
 
   // Decode sound
   int rid = enigma::sound_allocate();
   bool fail = enigma::sound_add_from_buffer(rid,fdata,flen);
-  delete fdata;
-  
+  delete [] fdata;
+
   if (fail)
     return -1;
   return rid;
 }
 
 bool sound_replace(int sound, string fname, int kind, bool preload)
 {
-	return false;
+  if (sound >= 0 && sound < sound_resources.size() && sound_resources[sound]) {
+    get_sound(snd,sound,false);
+    delete snd;
+  }
+
+  // Open sound
+  size_t flen = 0;
+  char *fdata = enigma::read_all_bytes(fname, flen);
+  if (!fdata) return -1;
+
+  // Decode sound
+  bool fail = enigma::sound_add_from_buffer(sound,fdata,flen);
+  delete [] fdata;
+  return fail;
 }
 
 void sound_3d_set_sound_cone(int sound, float x, float y, float z, double anglein, double angleout, long voloutside) {
+  get_soundv(snd, sound);
+
+  // query for the 3d buffer interface
+  IDirectSound3DBuffer8* sound3DBuffer8 = 0;
+  snd->soundBuffer->QueryInterface(IID_IDirectSound3DBuffer, (void**) &sound3DBuffer8);
+
+  sound3DBuffer8->SetConeOrientation(x, y, z, DS3D_IMMEDIATE);
+  sound3DBuffer8->SetConeAngles(anglein, angleout, DS3D_IMMEDIATE);
+  sound3DBuffer8->SetConeOutsideVolume(voloutside, DS3D_IMMEDIATE);
 }
 
 void sound_3d_set_sound_distance(int sound, float mindist, float maxdist) {
+  get_soundv(snd, sound);
+
+  // query for the 3d buffer interface
+  IDirectSound3DBuffer8* sound3DBuffer8 = 0;
+  snd->soundBuffer->QueryInterface(IID_IDirectSound3DBuffer, (void**) &sound3DBuffer8);
+
+  sound3DBuffer8->SetMinDistance(mindist, DS3D_IMMEDIATE);
+  sound3DBuffer8->SetMaxDistance(maxdist, DS3D_IMMEDIATE);
 }
 
 void sound_3d_set_sound_position(int sound, float x, float y, float z) {
+  get_soundv(snd, sound);
+
+  // query for the 3d buffer interface
+  IDirectSound3DBuffer8* sound3DBuffer8 = 0;
+  snd->soundBuffer->QueryInterface(IID_IDirectSound3DBuffer, (void**) &sound3DBuffer8);
+
+  sound3DBuffer8->SetPosition(x, y, z, DS3D_IMMEDIATE);
 }
 
 void sound_3d_set_sound_velocity(int sound, float x, float y, float z) {
-}
+  get_soundv(snd, sound);
 
-void sound_effect_chorus(int sound, float wetdry, float depth, float feedback, float frequency, long wave, float delay, long phase) {
-	/*
-	DSFXChorus pcDsFxChorus;
-	IDirectSoundFXChorus::SetAllParameters(pcDsFxChorus);
-	*/
-}
+  // query for the 3d buffer interface
+  IDirectSound3DBuffer8* sound3DBuffer8 = 0;
+  snd->soundBuffer->QueryInterface(IID_IDirectSound3DBuffer, (void**) &sound3DBuffer8);
 
-void sound_effect_compressor(int sound, float gain, float attack, float release, float threshold, float ratio, float delay) {
+  sound3DBuffer8->SetVelocity(x, y, z, DS3D_IMMEDIATE);
 }
 
-void sound_effect_echo(int sound, float wetdry, float feedback, float leftdelay, float rightdelay, long pandelay) {
+void sound_effect_chorus(int sound, float wetdry, float depth, float feedback, float frequency, long wave, float delay, long phase) {
+  /*
```

Comment:
Why with all the commented code? Is this why you were complaining about MinGW headers being outdated?

Reply:
Yep. Otherwise I could finish all of those functions, they just need one line to pass the effectParams structs to the effect interface obtained with QueryInterface.